sql-plugin/src/main/scala/com/nvidia/spark/rapids/PrioritySemaphore.scala

```diff
@@ -59,7 +59,9 @@ class PrioritySemaphore[T](val maxPermits: Int)(implicit ordering: Ordering[T])
     try {
       val condition = lock.newCondition()
       while (!canAcquire(numPermits)) {
-        waitingQueue.enqueue(ThreadInfo(priority, condition))
+        if (!waitingQueue.exists(_.condition == condition)) {
```

Comment:
I opted for a slightly different method that will also handle the case if we `release` from another thread (and dequeue this `ThreadInfo`) and then want to requeue.

Reply:
Not a fan of the linear scan here.  I think a better fix would be have `release` not try to wakeup threads that we know can't wakeup.  We can peek at the queue and see if it's worth waking up that thread based on the permits it's requesting and the permits available after releasing.

For extra sanity checking, @revans2 noted that we could put a boolean in `ThreadInfo` indicating it was dequeued (or "woken up") and then the waiting thread can trivially know whether to enqueue or not on a subsequent wait if it still can't acquire after wakeup.

Reply:
I didn't understand the last part exactly but I pushed a change which I think is in the same spirit.